src/XRP/xrp-payment.ts

```diff
@@ -0,0 +1,84 @@
+import { Payment } from '../generated/web/org/xrpl/rpc/v1/transaction_pb'
+import XRPCurrencyAmount from './xrp-currency-amount'
+import XRPPath from './xrp-path'
+
+// TODO(amiecorso): Modify this object to use X-Address format.
+export default class XRPPayment {
+  public static from(payment: Payment): XRPPayment | undefined {
+    let amount
+    const paymentAmountValue = payment.getAmount()?.getValue()
+    if (paymentAmountValue) {
+      amount = XRPCurrencyAmount.from(paymentAmountValue)
+      if (!amount) {
+        return undefined // amount is required
+      }
+    } else {
+      return undefined
+    }
+
+    const destination = payment
+      .getDestination()
+      ?.getValue()
+      ?.getAddress()
+    if (!destination) {
+      return undefined // destination is required
+    }
+
+    const destinationTag = payment.getDestinationTag()?.getValue()
+
+    // If the deliverMin field is set, it must be able to be transformed into a XRPCurrencyAmount.
+    let deliverMin
+    if (payment.hasDeliverMin()) {
+      const paymentDeliverMinValue = payment.getDeliverMin()?.getValue()
+      if (paymentDeliverMinValue) {
+        deliverMin = XRPCurrencyAmount.from(paymentDeliverMinValue)
+        if (!deliverMin) {
+          return undefined
+        }
+      }
+    } else {
+      deliverMin = undefined
+    }
+
+    const invoiceID = payment.getInvoiceId()?.getValue_asU8()
+
+    const paths =
+      payment.getPathsList()?.length > 0
+        ? payment.getPathsList().map((path) => XRPPath.from(path))
+        : undefined
+
+    // If the sendMax field is set, it must be able to be transformed into a XRPCurrencyAmount.
+    let sendMax
+    if (payment.hasSendMax()) {
+      const paymentSendMaxValue = payment.getSendMax()?.getValue()
+      if (paymentSendMaxValue) {
+        sendMax = XRPCurrencyAmount.from(paymentSendMaxValue)
+        if (!sendMax) {
+          return undefined
+        }
+      } else {
+        sendMax = undefined
+      }
+    }
+
+    return new XRPPayment(
+      amount,
+      destination,
+      destinationTag,
+      deliverMin,
+      invoiceID,
+      paths,
+      sendMax,
+    )
+  }
+
+  private constructor(
+    readonly amount?: XRPCurrencyAmount,
+    readonly destination?: string,
+    readonly destinationTag?: number,
+    readonly deliverMin?: XRPCurrencyAmount,
+    readonly invoiceID?: Uint8Array,
+    readonly paths?: (XRPPath | undefined)[],
```

Comment:
What is this type? An array of XRPPath or undefined? 

I would think that this is a `Array<XRPPath>|undefined`. Is this just alternative syntax? If so, I've generally been using `Array<Foo>` syntax in Xpring-JS, would you mind changing for consistency.

Reply:
So the issue with this is here: https://github.com/xpring-eng/Xpring-JS/blob/fa65158ff80240c92ac862702b52cd90fdb90736/src/XRP/xrp-payment.ts#L47  where we use the map function to convert an array of protobuf paths to an array of XRPPaths... since `XRPPath.from(...)` can return `undefined`, then it's possible that the array contains a mixture of `XRPPath` and `undefined` values.  (In practice, it never should, but the type system is obviously behaving this way).  I'm not sure how else I would go about that except perhaps to not use the map function?  And instead get rid of a value if it's `undefined` instead of ever allowing it to appear in the array?  The possibility of `undefined` is worse than coronavirus.

Reply:
and to answer your question specifically, I don't think this is alternative syntax, I think that what I have `(XRPPath | undefined)[]` means an array of mixed values? v.s. a single value that's not an array at all but just undefined.

Reply:
I think it is alternative syntax. You should be able to write:
```
Array<XRPPath | undefined>
```

right?

Reply:
> I'm not sure how else I would go about that except perhaps to not use the map function? And instead get rid of a value if it's undefined instead of ever allowing it to appear in the array? The possibility of undefined is worse than coronavirus.

I'm not sure where the relevant `.map()` call is, but you could do a `.filter()` afterwards to filter the array to truthy values.

Reply:
So I think it's alternative syntax for ```Array<XRPPath | undefined>``` but not for ```Array<XRPPath> | undefined```, which is what I thought Keefer was asking.  Should I re-write as ```Array<XRPPath | undefined>```  stylistically?

Reply:
oh, yes